parser.cs

```diff
@@ -130,15 +131,17 @@ public class Parser {
                 __symbolTableStack.Push(new SymbolTable(
                     programName, 0, 0, 0, new List<Entry>(),
                     null));
-                __analyzer.genInit();
+                progLabel = "L" + LabelMaker.genLabel();
+                __analyzer.genBr(progLabel);
                 break;
             default:
                 error(new List<TOKENS>{TOKENS.PROGRAM});
                 break;
         }
+        return progLabel;
     }
 
-    private void block() {
+    private void block(string startLabel, bool progBlock = false) {
```

Comment:
This is unnecessary, we don't need to know that we are in progBlock. If you really want to know what to do in the semantic analyzer at genEnd, just look at the nesting level. Level 0 = main, so hlt. Level 1+ = not main, so ret.